src/olympia/shelves/forms.py

```diff
@@ -0,0 +1,56 @@
+import requests
+
+from rest_framework.reverse import reverse as drf_reverse
+
+from django import forms
+from django.conf import settings
+
+from olympia.shelves.models import Shelf
+
+
+class ShelfForm(forms.ModelForm):
+    class Meta:
+        model = Shelf
+        fields = ('title', 'shelf_type', 'criteria',
+                  'footer_text', 'footer_pathname',)
+
+    def clean(self):
+        data = self.cleaned_data
+        criteria = data['criteria']
+        baseUrl = settings.EXTERNAL_SITE_URL
+
+        if data['shelf_type'] in ('extension', 'search', 'theme'):
+            api = drf_reverse('v4:addon-search')
+        elif data['shelf_type'] == 'categories':
+            api = drf_reverse('v4:category-list')
+        elif data['shelf_type'] == 'collections':
+            api = drf_reverse('v4:collection-list')
+        elif data['shelf_type'] == 'recommendations':
+            api = drf_reverse('v4:addon-recommendations')
+
+        url = baseUrl + api + criteria
+
+        try:
+            response = requests.get(url)
+            self.results = response.json()
+
+            if response.status_code == 404:
+                raise forms.ValidationError('Check criteria - No data found')
+            if response.status_code != 200:
+                raise forms.ValidationError(
+                    'Check criteria - %s' % self.results[0])
+
+            # Value of results is either dict or list depending on the endpoint
+            if 'count' in self.results:
+                if self.results.get('count', 0) == 0:
+                    raise forms.ValidationError(
+                        'Check criteria parameters - e.g., "type"')
+            elif len(self.results) == 0:
+                raise forms.ValidationError('Check criteria - No data found')
+        except requests.exceptions.ConnectionError as err:
+            print('Error connecting:', err)
+
+    def save(self, commit=True):
+        instance = super().save(commit=False)
+        instance.results = self.results
+        instance.save()
```

Comment:
> this is all duplicated with the code in forms, and would mean two identical requests, so we don't want that. I'm guessing this is just because you need the response to include the count in the messages.
> 
> If so, lets just hack around that - store the response on the form instance `self.response = response.json()`, then add it to the Shelf instance (in the form's save method), then it'll be available as `obj.response` in `save_model`.

I'm not following how `ShelfAdmin.save_model` gets the add-ons count from the form's `clean` method if the `save_model`'s duplicate call to the API is eliminated...

Reply:
this `save` method is supposed to return `instance` (I missed that).

So, django admin does it (roughly) in this order:
- form.valid(), which calls form.clean()
- Admin.save_form(), which calls form.save() - that should return the new/update instance, but _doesn't_ save it to the database (if commit=False, which is what it passes in Admin.save_form)
- Admin.save_model(), which calls obj.save() and actually saves the data to the database.

https://github.com/django/django/blob/2.2.14/django/contrib/admin/options.py#L1745:L1773  (it's a bit overly complicated, but if you're interested, take a look)